shotover-proxy/examples/windsock/redis.rs

```diff
@@ -0,0 +1,293 @@
+        let perf = if flamegraph {
+            if let Some(shotover) = &shotover {
+                Some(Perf::new(shotover.child.as_ref().unwrap().id().unwrap()))
+            } else {
+                todo!()
+            }
+        } else {
+            None
+        };
+
+        let mut config = RedisConfig::from_url(address).unwrap();
+        if let Encryption::Tls = self.encryption {
+            let private_key = load_private_key("tests/test-configs/redis-tls/certs/localhost.key");
+            let certs = load_certs("tests/test-configs/redis-tls/certs/localhost.crt");
+            config.tls = Some(
+                ClientConfig::builder()
+                    .with_safe_defaults()
+                    .with_root_certificates(load_ca(
+                        "tests/test-configs/redis-tls/certs/localhost_CA.crt",
+                    ))
+                    .with_single_cert(certs, private_key)
+                    .unwrap()
+                    .into(),
+            );
+        }
+        let client = Arc::new(RedisClient::new(config, None, None));
+
+        // connect to the server, returning a handle to the task that drives the connection
+        let shutdown_handle = client.connect();
+        client.wait_for_connect().await.unwrap();
+
+        if let RedisOperation::Get = self.operation {
+            let _: () = client.set("foo", 42, None, None, false).await.unwrap();
+        }
+
+        let tasks = BenchTaskRedis {
+            client: client.clone(),
+            operation: self.operation,
+        }
+        .spawn_tasks(reporter.clone(), operations_per_second)
+        .await;
+
+        // warm up and then start
+        tokio::time::sleep(Duration::from_secs(1)).await;
+        reporter.send(Report::Start).unwrap();
+        let start = Instant::now();
+
+        for _ in 0..runtime_seconds {
+            let second = Instant::now();
+            tokio::time::sleep(Duration::from_secs(1)).await;
+            reporter
+                .send(Report::SecondPassed(second.elapsed()))
+                .unwrap();
+        }
+
+        reporter.send(Report::FinishedIn(start.elapsed())).unwrap();
+
+        // make sure the tasks complete before we drop the database they are connecting to
+        for task in tasks {
+            task.await.unwrap();
+        }
+
+        client.quit().await.unwrap();
+        shutdown_handle.await.unwrap().unwrap();
+
+        if let Some(shotover) = shotover {
+            shotover
+                .shutdown_and_then_consume_events(&[EventMatcher::new()
+                    .with_level(Level::Error)
+                    .with_message("encountered error in redis stream: Io(Kind(UnexpectedEof))")
+                    .with_target("shotover::transforms::redis::sink_single")
+                    .with_count(Count::Any)])
+                .await;
+        }
+
+        if let Some(perf) = perf {
+            perf.flamegraph();
+        }
+    }
+}
+
+fn load_certs(path: &str) -> Vec<Certificate> {
```

Comment:
What's the reason for these `inner` functions?

Reply:
avoids duplicating the filename in the error handling

Reply:
If we end up adding more drivers that also expose rustls like this then we can pull this logic into a test-helpers module.